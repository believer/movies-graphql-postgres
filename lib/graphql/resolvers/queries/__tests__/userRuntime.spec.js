const userRuntime = require('../userRuntime')
const { verifyToken } = require('../../../../services/token')

jest.mock('../../../../services/token', () => ({
  verifyToken: jest.fn().mockImplementation(() =>
    Promise.resolve({
      id: 2,
    })
  ),
}))

describe('queries/userRuntime', () => {
  let db

  beforeEach(() => {
    db = {
      query: jest.fn().mockReturnValue([{ hours: 10 }, { minutes: 60 }]),
    }
  })

  it('should verify token', () => {
    userRuntime(null, {}, { db, token: 'test' })

    expect(verifyToken).toHaveBeenCalledWith('test')
  })

  it('should get how many movies a for a specific role', async () => {
    await userRuntime(null, {}, { db })

    expect(db.query.mock.calls[0][0]).toMatchSnapshot()
  })

  it('should return database data', async () => {
    const response = await userRuntime(null, {}, { db })

    expect(response).toEqual({
      total_with_rewatches: { hours: 10 },
      total: { minutes: 60 },
    })
  })

  it('should handle errors', async () => {
    db.query.mockImplementation(() => Promise.reject('nope'))

    try {
      await userRuntime(null, {}, { db })
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
