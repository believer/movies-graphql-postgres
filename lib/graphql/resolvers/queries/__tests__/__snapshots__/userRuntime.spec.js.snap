// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries/userRuntime should get how many movies a for a specific role 1`] = `
"
      SELECT
        all_views.minutes,
        all_views.minutes / 60 AS hours,
        all_views.minutes / (60 * 24) AS days,
        all_views.minutes / (60 * 24 * 365)::float AS years
      FROM (
        SELECT SUM(m.runtime) as minutes
        FROM views AS v 
        INNER JOIN movies AS m ON m.id = v.movie_id 
        WHERE v.user_id = 2
      ) AS all_views;

      SELECT
        all_unique_views.minutes,
        all_unique_views.minutes / 60 AS hours,
        all_unique_views.minutes / (60 * 24) AS days,
        all_unique_views.minutes / (60 * 24 * 365)::float AS years
      FROM (
        SELECT SUM(unique_views.runtime) as minutes
        FROM (
          SELECT 
            distinct m.id, 
            m.runtime 
          FROM views AS V 
          INNER JOIN movies AS m ON m.id = v.movie_id 
          WHERE v.user_id = 2
        ) as unique_views
      ) AS all_unique_views;
    "
`;
